Improves pom dynamic revision conversion

eed3si9n opened the pull request:
sbt/sbt@5486daf7006462be13b82d6ae225e816f038c9ff fixed #1275, but it didn't close the issue.
Not realizing that, I originally set out to fix #1275, and mostly did code cleanup and minor enhancement.
- `+` now translates to `[0,)`
- `foo+` now errors.
## notes
- http://ant.apache.org/ivy/history/2.3.0/ivyfile/dependency.html
- http://maven.apache.org/enforcer/enforcer-rules/versionRanges.html

/review @havocp

eed3si9n committed:
Fixes #1275. Fixes pom dynamic revision conversion

Ivy and pom uses slightly different notation for version range and
dynamic revision.
This change fixes the dynamic revisions involving “+”.
First, when a revision like “1.1+” is found, it will now be treated as
“1.+”.
Next, when it finds a revision like “1+” is found, it will be treated
as “+”.
The conversion of “+” is hardcoded to be “[0,)”.

havocp commented:
I lack all background knowledge here other than reading the bug and links above. http://ant.apache.org/ivy/history/2.3.0/ivyfile/dependency.html does not seem to say what `1.1+` means, it says what `1.1.+` means? If `1.1+` is nonsense are we only trying to make sense of it for back compat reasons? Can we print some kind of warning "you are talking nonsense?" Does Ivy do something with `1.1+` and if so what?

havocp commented on ivy/src/main/scala/sbt/MakePom.scala:
```diff
@@ -209,20 +210,24 @@ class MakePom(val log: Logger) {
-      } else revision
+      revision match {
+        case PlusPattern(base, tail) => plusRange(base)
+        // There's no direct translation for a dynamic revision like "1+".
```
you have the "stuff is busted" comment here, but I think the above PlusPattern matches the busted case `1.1+` right? I was a little confused expecting that the comment applied only to what was below it

havocp commented on ivy/src/main/scala/sbt/MakePom.scala:
```diff
@@ -209,20 +210,24 @@ class MakePom(val log: Logger) {
+        // So if someone wanted (1.0 <= x), then it should be "[1.0)" or "1.+".
+        // Technically speaking, "1+" should convert to "LATEST",
+        // but that seems to be deprecated now, so picking the minimum version "0".
+        case rev if rev endsWith "+" => "[0,)"
```
this will accept `total gobbledygook blaah blashdflh sdfh sdlj 4;l5kj423;lk5j4123;kj +` right? I guess versions can have alpha chars so we just have to live with that? Or is there some kind of other filtering going on before we get here

havocp commented on ivy/src/test/scala/MakePomSpec.scala:
```diff
@@ -0,0 +1,60 @@
+    ${convertTo("1.1+", "[1,2)")}
+
+  1+ should
+    ${convertTo("1+", "[0,)")}
```
other cases I can think of: just `+` by itself, `somejunk +`, `wedontunderstandthis`

eed3si9n commented:
Since sbt uses Ivy for resolution, I was able to experiment with various version numbers involving "+".

```
libraryDependency                          result
"commons-io" % "commons-io" % "1.1+"       commons-io-1.1.jar
"commons-io" % "commons-io" % "1+"         commons-io-1.4.jar
"commons-io" % "commons-io" % "+"          commons-io-20030203.000550.jar
```

As far as Ivy's concerned `somejunk +` is also a valid revision. Maven Central just happens to not have an artifact published with the version number:

```
[info] Resolving commons-io#commons-io;somejunk + ...
[warn]  module not found: commons-io#commons-io;somejunk +
[warn] ==== local: tried
[warn]   /Users/eugene/.ivy2/local/commons-io/commons-io/[revision]/ivys/ivy.xml
[warn] ==== public: tried
[warn]   http://repo1.maven.org/maven2/commons-io/commons-io/[revision]/commons-io-[revision].pom
[warn]   [20030203.000550, 2.4, 2.3, 2.2, 2.1, 2.0.1, 2.0, 1.4, 1.3.2, 1.3.1, 1.3, 1.2, 1.1, 1.0, 0.1]
```

This means that my assumption that "1.1+" should be treated as "1.+" is incorrect. It should in fact be treated as "1.1.+".

I carried over `revision endsWith "+"` from the old code, but I could reject `somejunk +` to be "Not translatable to Maven POM."

eed3si9n committed:
Based on Ivy's behavior, bringing back approximation

I assumed 1.1+ should be treated as 1.+, but it seems like Ivy treats
it more as “any version that starts with 1.1” including 1.10.
Josh’s original implementation approximates this by making ranges for
multiple digits, 1.1~1.2, 1.10~1.20, etc.

eed3si9n commented:
Added a new commit, but realized that the issue that I set out to fix has already been fixed. This PR ended up being code cleanup and minor enhancement.

jsuereth commented:
LGTM
